aws-lc-rs/src/rsa.rs

```diff
@@ -507,69 +546,99 @@ impl RsaParameters {
+    let mut cbs = cbs::build_CBS(private_key);
+
+    let rsa = DetachableLcPtr::new(RSA_parse_private_key(&mut cbs))?;
+
+    let pkey = LcPtr::new(EVP_PKEY_new())?;
+
+    if 1 != EVP_PKEY_assign_RSA(*pkey, *rsa) {
+        return Err(KeyRejected::unexpected_error());
+    }
+
+    rsa.detach();
 
-    let rsa =
-        LcPtr::new(RSA_parse_private_key(&mut cbs)).map_err(|_| KeyRejected::invalid_encoding())?;
-    Ok(rsa)
+    Ok(pkey)
 }
 
 #[inline]
 #[allow(non_snake_case)]
 fn verify_RSA(
     algorithm: &'static digest::Algorithm,
     padding: &'static RsaPadding,
-    public_key: &LcPtr<*mut RSA>,
+    public_key: &LcPtr<*mut EVP_PKEY>,
     msg: &[u8],
     signature: &[u8],
     allowed_bit_size: &RangeInclusive<u32>,
 ) -> Result<(), Unspecified> {
-    unsafe {
-        let n = ConstPointer::new(RSA_get0_n(**public_key))?;
-        let n_bits = n.num_bits();
-        if !allowed_bit_size.contains(&n_bits) {
-            return Err(Unspecified);
-        }
+    let rsa = DetachableLcPtr::new(unsafe { EVP_PKEY_get0_RSA(**public_key) })?;
+    let n = ConstPointer::new(unsafe { RSA_get0_n(rsa.detach()) })?;
+    let n_bits = n.num_bits();
+    if !allowed_bit_size.contains(&n_bits) {
+        return Err(Unspecified);
+    }
 
-        let digest = digest::digest(algorithm, msg);
-        let digest = digest.as_ref();
-
-        let result = match padding {
-            RsaPadding::RSA_PKCS1_PADDING => RSA_verify(
-                algorithm.hash_nid,
-                digest.as_ptr(),
-                digest.len(),
-                signature.as_ptr(),
-                signature.len(),
-                **public_key,
-            ),
-            RsaPadding::RSA_PKCS1_PSS_PADDING => RSA_verify_pss_mgf1(
-                **public_key,
-                digest.as_ptr(),
-                digest.len(),
-                *match_digest_type(&algorithm.id),
-                null(),
-                -1,
-                signature.as_ptr(),
-                signature.len(),
-            ),
-        };
+    let mut md_ctx = digest::digest_ctx::DigestContext::new_uninit();
+    let digest = digest::match_digest_type(&algorithm.id);
+
+    let mut pctx = MaybeUninit::<*mut EVP_PKEY_CTX>::uninit();
+
+    if 1 != unsafe {
+        EVP_DigestVerifyInit(
+            md_ctx.as_mut_ptr(),
+            pctx.as_mut_ptr(),
+            *digest,
+            null_mut(),
+            **public_key,
+        )
+    } {
+        return Err(Unspecified);
+    }
+
+    // This is allocation is managed by aws-lc so need for us to wrap it for freeing.
+    let pctx = unsafe { pctx.assume_init() };
 
-        if result != 1 {
+    if let RsaPadding::RSA_PKCS1_PSS_PADDING = padding {
+        if 1 != unsafe { EVP_PKEY_CTX_set_rsa_padding(pctx, RSA_PKCS1_PSS_PADDING) } {
             return Err(Unspecified);
-        }
-        Ok(())
+        };
+        if 1 != unsafe { EVP_PKEY_CTX_set_rsa_pss_saltlen(pctx, -1) } {
```

Comment:
```suggestion
        if 1 != unsafe { EVP_PKEY_CTX_set_rsa_pss_saltlen(pctx, RSA_PSS_SALTLEN_DIGEST) } {
```
This seems to be using similar calls to the padding functions in `sign`. Could we abstract them into another function?